### PR TITLE
feat Progressive image loading

### PR DESCRIPTION
**Pull request by Harshith46**
# Description

we are rendering a smaller size image on the initial load and then loading a higher quality of the image after few seconds. Doing this will improve page performance as we render a smaller size image initially. this size, time to load high quality image and other params are configurable from bk

Fixes: https://github.com/quintype/ace-planning/issues/419

Documentation: will taken in diff pr

## Type of change

Please delete options that are not relevant.

- [ ] New feature (non-breaking change which adds functionality)
- [ ] This change requires a documentation update

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [ ] tested locally and on staging for backend compatibity
- [ ] check the improved scores on page speed insights

# Checklist:

- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

**Review by Jeevan-Kishore — CHANGES_REQUESTED** [review]
There were issues with huge DOM, multiple / duplicate API requests. Are those sorted?

File: app/isomorphic/components/atoms/card-image/card-image.m.css [hunk 1]
```diff
@@ -0,0 +1,7 @@
+.card-image {
+  width: 100%;
+}
+
+.placeholder {
+  background-color: lightgray;
```
Comment (Deokumarjnu) [inline app/isomorphic/components/atoms/card-image/card-image.m.css:6]: is this color defined anywhere?

File: app/isomorphic/components/atoms/card-image/index.js [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+  const imagePerfObj = progressiveImageConfig.is_enable
+    ? progressiveImageConfig.initial_load
+    : { size: "25vw", blur: 0 };
+  const [perfObj, setPerfObj] = useState(imagePerfObj);
+  const customStyleName = pageType !== "story-page" && !story["hero-image-s3-key"] ? "placeholder" : "";
+
+  useEffect(() => {
+    if (isInitRow) {
+      // if we need progressive loading for the images in the page, this condition can be removed
+      setTimeout(() => {
```
[the remark this answers is not in the document]
Comment (Harshith46) [reply app/isomorphic/components/atoms/card-image/index.js:22]: sure

File: app/isomorphic/components/atoms/card-image/card-image.m.css (hunk 1 above)
Comment (Harshith46) [reply app/isomorphic/components/atoms/card-image/card-image.m.css:6]: will add variable name

File: app/isomorphic/components/atoms/card-image/index.js [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+    get(state, ["qt", "config", "publisher-attributes", "progressive_image"], {})
+  );
+
+  const imagePerfObj =
+    progressiveImageConfig.is_enable && isInitRow ? progressiveImageConfig.initial_load : { size: "25vw", blur: 0 };
+  const [perfObj, setPerfObj] = useState(imagePerfObj);
+  const customStyleName = pageType !== "story-page" && !story["hero-image-s3-key"] ? "placeholder" : "";
+
+  useEffect(() => {
+    if (isInitRow && progressiveImageConfig.is_enable) {
```
Comment (Harshith46) [inline app/isomorphic/components/atoms/card-image/index.js:20]: mention about the isInitrow condition

**Review by ags1773 — APPROVED** [review]
Looks good, provided there's a more detailed explanation for `isInitRow`

**Review by Deokumarjnu — APPROVED** [review]

**Review by Athira001 — APPROVED** [review]

File: app/isomorphic/components/atoms/card-image/index.js [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+      // if we need progressive loading for the images in the page, isInitRow condition can be removed
+      const timer = () => {
```
Comment (Jeevan-Kishore) [inline app/isomorphic/components/atoms/card-image/index.js:22]: clearly this wont work. Please take care on what's returned and what isn't from a function

File: app/isomorphic/components/atoms/card-image/index.js [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+        setTimeout(() => {
```
Comment (Jeevan-Kishore) [inline app/isomorphic/components/atoms/card-image/index.js:23]: carve out function definition and only call in useEffect

File: app/isomorphic/components/atoms/card-image/index.js [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+  const getImageSize = () => {
+    if (pageType === "story-page") {
+      return perfObj.story_page_size;
+    }
+    return perfObj.generic_size;
+  };
+
+  return (
+    <figure className="qt-image-16x9" styleName={`card-image ${customStyleName}`}>
+      {story["hero-image-s3-key"] && (
```
Comment (Jeevan-Kishore) [inline app/isomorphic/components/atoms/card-image/index.js:42]:
```suggestion
      !story["hero-image-s3-key"] return null;
```

Comment (Jeevan-Kishore) [reply app/isomorphic/components/atoms/card-image/index.js:42]: Use a function if need be

**Review by Jeevan-Kishore — CHANGES_REQUESTED** [review]
Can i find documentation of this somewhere?

File: app/isomorphic/components/atoms/card-image/index.js [hunk 7]
```diff
@@ -0,0 +1,76 @@
+import React, { useEffect, useState } from "react";
+import { get } from "lodash";
```
Comment (Jeevan-Kishore) [inline app/isomorphic/components/atoms/card-image/index.js:2]:
```suggestion
import get from "lodash/get";
```

File: app/isomorphic/components/atoms/card-image/index.js [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+  const imagePerfObj =
+    progressiveImageConfig.is_enable && isInitRow ? progressiveImageConfig.initial_load : { size: "25vw", blur: 0 };
+  const [perfObj, setPerfObj] = useState(imagePerfObj);
+  const customStyleName = pageType !== "story-page" && !story["hero-image-s3-key"] ? "placeholder" : "";
+
+  useEffect(() => {
+    if (isInitRow && progressiveImageConfig.is_enable) {
+      // if we need progressive loading for the images in the page, isInitRow condition can be removed
+      const timer = setTimeout(() => {
+        setPerfObj(progressiveImageConfig.subsequent_load);
```
Comment (Jeevan-Kishore) [inline app/isomorphic/components/atoms/card-image/index.js:23]: clear timer once, not on unmount

**Review by Jeevan-Kishore — APPROVED** [review]
Approved with comments

**Comment by Harshith46** [comment]
docs: https://github.com/quintype/malibu/pull/772

**Pull request merged**